go/ephemeral/team_ek.go

```diff
@@ -0,0 +1,413 @@
+	defer g.CTrace(ctx, "boxTeamEKForMembers", func() error { return err })()
+
+	membersMetadata, err := activeMemberEKMetadata(ctx, g, teamID, merkleRoot)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	myUID := g.Env.GetUID()
+	for uid, memberMetadata := range membersMetadata {
+		recipientKey, err := libkb.ImportKeypairFromKID(memberMetadata.Kid)
+		if err != nil {
+			return nil, nil, err
+		}
+		// Encrypting with a nil sender means we'll generate a random sender private key.
+		box, err := recipientKey.EncryptToString(seed[:], nil)
+		if err != nil {
+			return nil, nil, err
+		}
+		boxMetadata := TeamEKBoxMetadata{
+			RecipientUID:        uid,
+			RecipientGeneration: memberMetadata.Generation,
+			Box:                 box,
+		}
+		boxes = append(boxes, boxMetadata)
+
+		if uid == myUID {
+			myTeamEKBoxed = &keybase1.TeamEkBoxed{
+				Box:              box,
+				UserEkGeneration: memberMetadata.Generation,
+				Metadata:         teamMetadata,
+			}
+		}
+	}
+	return boxes, myTeamEKBoxed, err
+}
+
+type teamEKStatementResponse struct {
+	Sig *string `json:"sig"`
+}
+
+// Returns nil if the team has never published a teamEK. If the team has
+// published a teamEK, but has since rolled their PTK without publishing a new
+// one, this function will also return nil and log a warning. This is a
+// transitional thing, and eventually when all "reasonably up to date" clients
+// in the wild have EK support, we will make that case an error.
+func fetchTeamEKStatement(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID) (statement *keybase1.TeamEkStatement, err error) {
+	defer g.CTrace(ctx, "fetchTeamEKStatement", func() error { return err })()
+
+	apiArg := libkb.APIArg{
+		Endpoint:    "team/team_ek",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		NetContext:  ctx,
+		Args: libkb.HTTPArgs{
+			"team_id": libkb.S{Val: string(teamID)},
+		},
+	}
+	res, err := g.GetAPI().Get(apiArg)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedResponse := teamEKStatementResponse{}
+	err = res.Body.UnmarshalAgain(&parsedResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the result field in the response is null, the server is saying that
+	// the team has never published a teamEKStatement, stale or otherwise.
+	if parsedResponse.Sig == nil {
+		g.Log.CDebugf(ctx, "team has no teamEKStatement at all")
+		return nil, nil
+	}
+
+	statement, wrongKID, err := verifySigWithLatestPTK(ctx, g, teamID, *parsedResponse.Sig)
+	// Check the wrongKID condition before checking the error, since an error
+	// is still returned in this case. TODO: Turn this warning into an error
+	// after EK support is sufficiently widespread.
+	if wrongKID {
+		g.Log.CWarningf(ctx, "It looks like you revoked a user without generating new ephemeral keys. Are you running an old version?")
```

Comment:
I think the condition here is that someone has rolled the PTK (CLKR'd the PTK? acronyms getting out of hand...) without issuing a new teamEK.